### ISSUE TITLE
mypy incorrectly reports incompatible return types for overloaded methods

### ISSUE THREAD

**Issue by Prometheus3375** [python/mypy#12733]
**Bug Report**

```python
@overload
def foo(cls: type[dict], mapping: SupportsKeysAndGetItem[K, V], /) -> dict[K, V]: ...
@overload
def foo(cls: type[dict], mapping: SupportsKeysAndGetItem[str, V], /, **kwargs: V) -> dict[str, V]: ...
```
Return types are compatible.

```python
class FrozendictBase(Generic[K, V]):
    @overload
    def __new__(
            cls,
            mapping: SupportsKeysAndGetItem[K, V],
            /,
            ) -> 'FrozendictBase[K, V]': ...

    @overload
    def __new__(
            cls,
            mapping: SupportsKeysAndGetItem[str, V],
            /,
            **kwargs: V,
            ) -> 'FrozendictBase[str, V]': ...
```
Return types are also compatible, but mypy reports an error.

**To Reproduce**

1. Download [file.txt](https://github.com/python/mypy/files/8633239/file.txt) and save as `sandbox.py`.
3. Run `mypy sandbox.py`.

**Expected Behavior**
```
sandbox.py:8: error: Invariant type variable "V" used in protocol where covariant one is expected
```
This is due to #5775.

**Actual Behavior**

```
sandbox.py:8: error: Invariant type variable "V" used in protocol where covariant one is expected
sandbox.py:39: error: Overloaded function signatures 3 and 4 overlap with incompatible return types
sandbox.py:54: error: Overloaded function signatures 5 and 6 overlap with incompatible return types
```

**Your Environment**
- Mypy version used: mypy 0.950
- Mypy command-line flags:
- Mypy configuration options from `mypy.ini` (and other config files):
- Python version used: Python 3.10.4
- Operating system and version: Windows 10

**Comment by hauntsaninja** [comment]
I thiiiiink mypy is correct here, although it's certainly not intuitive.

The tricky thing is dict is invariant in its key type. So if you pass `SupportsKeysAndGetItem[StrSubclass, int]` and no extra kwargs, both overloads would match, but overload 1 returns you a `dict[StrSubclass, int]` and overload 2 returns you a `dict[str, int]` and those are incompatible.

**Comment by Prometheus3375** [comment]
@hauntsaninja
1. mypy report an error only in FrozendictBase methods.
2. `SupportsKeysAndGetItem` is also invariant in both key and values, so `SupportsKeysAndGetItem[StrSubclass, int]` is not subclass of `SupportsKeysAndGetItem[str, int]`; thus, overload 1 is not macthed.
